debian/dswm.1
.\" -*- nroff -*-
.\"
.\" Copyright (c) 2007-2008 Luca Capello <luca@pca.it>

.TH DSWM 1 "Jul 23, 2008"

.SH NAME
dswm \- a Common Lisp window manager

.SH SYNOPSIS
.B dswm

.SH DESCRIPTION
DSwm is a tiling, keyboard driven X11 Window Manager written
entirely in Common Lisp.

DSWM attempts to be customizable yet visually minimal.  There are
no window decorations, no icons, and no buttons.  It does have various
hooks to attach your personal customizations, and variables to tweak.

.SH FILES
.I ~/.dswmrc

.SH ENVIRONMENT
.B LISP
.IP
Specify which Common Lisp implementation should be used.  Default to
CLISP, other values are SBCL and CMUCL (all lowercase).  This can also
be configured in \fI~/.dswmrc\fP as "\fB;debian=lisp\fP".

.SH SEE ALSO
.BR clisp (1),
.BR sbcl (1),
.BR cmucl (1)
.P
The complete DSWM documentation is available under the info
subsystem, as well as the SLIME documentation.

.SH AUTHORS
DSWM was principally written by Shawn Betts <sabetts@vcn.bc.ca> and
is licensed under the terms of the GNU General Public License (GPL),
version 2 or later.

This manual page was written by Luca Capello <luca@pca.it> for the
Debian DSWM package and is licensed under the terms of the GNU
General Public License (GPL), version 2 or later.  Check
/usr/share/doc/dswm/copyright for more information.


NEWS
-*- outline -*-

This file documents user visible changes between versions of DSWM
* Changes since 0.1

** Rewrite build system with full autotools support
** add keys in ./configure script:
   --with-ql[=DIR] to support resolve dependencies with quicklisp
   --enable-dbus DBUS native support
   --enable-[gnome|cinnamon|kde|xfce|lxde] to run as window manager in most popular desktop environments
   --enable-debug to support lisp debug mode while build


* Changes since 0.0.5

** add remember-forget-reacll subsystem
*** add remember,forget,recall commands with options:
    group,screen,desktop,window,group-windows,screen-windows,all-windows,all

** add fselect-by-number command

** add keys by-name-p and by-number-p to select-group function
   (select-group screen query :by-name-p <t/nil> :by-number-p <t/nil>)

** add keys in ./configure script:
   --enable-shared                     - enables DSWM build as shared library + startup script
   --enable-static                     - enables DSWM build as static binary file
   --with-lisp-extra-options="options" - possibility to add extra options to lisp compiler
   --clisp-link			       - choose, with what CLISP kernel do you want to build DSWM
   --dswm-source-dir		       - directory with DSWM source (for --ehable-shared option)
   --with-info                         - also install info pages for DSWM

** add 'gmove-scratchpad' command (bound to PREF g s)

** add fprev command

** add 'edit-variable' command (alias to 'edit')

** add support for clisp and ccl

** add ENVIRONMENT variables support for 'browser', 'shell', 'editor' and 'terminal' commands
   *editor* -> $EDITOR -> 'emacs'
   *shell* -> $SHELL -> 'sh -c'
   *browser* -> $BROWSER -> 'conkeror'
   *terminal* -> $TERM -> 'xterm'

** add 'gconvert' command and 'convert-group' function

** add *event-processing-hook* hook

** add *quit-hook* hook

** add full window-switching support to floating group
*** add commands: 'float-focus-next' 'float-focus-previous' 'float-focus-other'
*** add keybindings: 'PREF n' 'PREF SPC' 'PREF C-SPC' 'PREF p' 'PREF o'

** add basic windows support, using cygwin (CLISP only)

** add key argument 'number' to add-group function
   (add-group screen name :number <number> :background <t/nil>)
** change iresize behavior to more intuitive

** replace configuration and data files path to XDG recommendations:
   system configfile: XDG_CONFIG_DIRS/dswm/dswm.lisp -> /etc/xdg/dswm/dswm.lisp
   user configfile: ~/.dswm -> XDG_CONFIG_HOME/dswm/init.lisp -> ~/.config/dswm/init.lisp
   system configdir: XDG_CONFIG_DIRS/dswm/ -> /etc/xdg/dswm/
   user configdir: XDG_CONFIG_HOME/dswm/ -> ~/.config/dswm/
   system datadir: XDG_DATA_DIRS/dswm/
   user datadir: XDG_DATA_HOME/dswm/
   system modules dir: <SYSTEM_DATA_DIR>/dswm/modules
   user modules dir: <USER_DATA_DIR>/dswm/modules -> ~/.local/share/dswm/modules
*** add functions '(conf-dir)', '(conf-dir-file <file> &optional <subdir>)'

** replace 'emacs' to 'editor' command
*** added *editor* variable

** replace commands:
   'show-window-properties' -> 'window-info' ('info' saved)
   'list-window-properties' -> 'window-full-info'

** remove commands:
   'groups' - use 'grouplist' or 'vgroups' instead
   'windows' - use 'windowlist' or 'vgroups' instead

** fix scratchpad number changing, when DSWM reloads

** fix black screen in fullscreen games

** fix bug in 'renumber' command (set <window number>+1 as real window number)

** fix bug in command 'repack-window-numbers'

** fix bug in command 'grun-new'

** fix bug with deleting last group

** fix bug with pull, when number of windows more, than mapping count

**

* Changes since 0.0.4

** fix error with desktop dump, when groups are empty

** fix error message on remember-windows-current-screen

** fix few minor bugs

** add autocompletion feature for 'bind' command

** add editing feature for some commands (ex. old group name by default, for 'grename' etc)

** add separate history for different commands

** add HEX format for color definitions
   ex.: (set-bg-color (make-color-hex "#000000")) also you can use (set-bg-color "red")

** add commands 'module-info', 'modules', 'load-file', 'loaded-modules'

** add variables *scratchpad-group-name*, *menu-window-gravity*

** add alias gselect-with-window -> gmove

** add functions
   'set-mode-line-{bg,fg,border}-color'
   'set-screen-mode-line-format'
   'set-screen-widget-line-format'
   'make-hex-color'
   'set-mode-line-{fg,bg,border}-color'
   'set-mode-line-border-width'
   'set-float-window-title-height'

** add function 'startup-only' to execute code only with startup (not when 'initrc' command called)

** add commands forget-desktop recall-desktop (for RFR model)

** change function 'data-dir'->'data-dir' with optional parameter 'subdir'

** change function data-dir-file->data-dir-file &key type subdir

** change default color scheme

** change modules format to asdf-compatible

** change commands to functions
   {remember,forget}-windows-current-{group,screen},version,ratwarp,ratrelwarp,ratclick,frame-windows

** rename mode-line variables
   *screen-mode-line-format*->*mode-line-format*
   *screen-widget-line-format*->*info-line-format*
   *screen-window-list-line-format*->*window-list-line-format*

** rename macro when-not-null->if-not-null

** remove commands 'groups', 'windows' (use vgroups and grouplist for visualisation and echo-groups for programming), 'version' (use help-short to see DSWM version)

** remove macros 'defkey(s)-{top,root} as non functional in standard complectation

** remove remember-group remember-screen as deprecated

* Changes since 0.0.3
** synchronize code with newer stumpwm code

** add hook frame-split-hook

** fix bug in 'run-in-terminal' command

** add functions: grun

** add commands: edit-variable add-modules-dir help-short

** modify dump-restore subsystem. New commands:
   remember-window remember-windows-current-{group,screen} remember-all-windows

src/dswm.lisp
;; Copyright (C) 2003-2008 Shawn Betts
;; Copyright (C) 2010-2012 Alexander aka CosmonauT Vynnyk
;;
;;  This file is part of dswm.
;;
;; dswm is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.

;; dswm is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this software; see the file COPYING.  If not, see
;; <http://www.gnu.org/licenses/>.

;; Commentary:
;;
;; Code:

(in-package :dswm)

(export '(cancel-timer
	  run-with-timer
	  dswm
	  startup-only
	  timer-p))


;;; Main

(defmacro startup-only (&rest body)
  "function, which runs code, just when DSWM starting.
Not when command loadrc runs, reinit runs after crash etc.
Useful for run programs on startup etc."
  `(setf *startup-only-code* (cons 'list ',body)))

(defun load-rc-file (&optional (catch-errors t) (reload nil))
  "Load the user's .dswm file or the system wide one if that
doesn't exist. Returns a values list: whether the file loaded (t if no
rc files exist), the error if it didn't, and the rc file that was
loaded. When CATCH-ERRORS is nil, errors are left to be handled further up. "
  (let* ((user-rc (file-exists-p (conf-dir-file "init" "lisp")))
	 (user-initrc (file-exists-p (merge-pathnames (user-homedir-pathname)
						      #p".dswm")))
	 (etc-rc (find-etc-file "dswm" "lisp"))
	 (rc (or user-initrc user-rc)))
    (progn
      (startup-only)
      (if etc-rc
	  (if catch-errors
	      (handler-case (load etc-rc)
		(error (c) (values nil (format nil "~a" c) etc-rc))
		(:no-error (&rest args) (declare (ignore args)) (values t nil etc-rc)))
	    (progn
	      (load etc-rc)
	      (values t nil etc-rc)))
	  (values t nil t))
      (if rc
	  (if catch-errors
	      (handler-case (load rc)
		(error (c) (values nil (format nil "~a" c) rc))
		(:no-error (&rest args) (declare (ignore args)) (values t nil rc)))
	      (progn
		(load rc)
		(values t nil rc)))
	  (values t nil nil))
      (when (not reload)
	(eval *startup-only-code*)))))

(defun error-handler (display error-key &rest key-vals &key asynchronous &allow-other-keys)
  "Handle X errors"
  (cond
    ;; ignore asynchronous window errors
    ((and asynchronous
          (find error-key '(xlib:window-error xlib:drawable-error xlib:match-error)))
     (dformat 4 "Ignoring error: ~s~%" error-key))
    ((eq error-key 'xlib:access-error)
     (write-line "Another window manager is running.")
     (throw :top-level :quit))
     ;; all other asynchronous errors are printed.
     (asynchronous
      (message "Caught Asynchronous X Error: ~s ~s" error-key key-vals))
     (t
      (apply 'error error-key :display display :error-key error-key key-vals))))

;;; Timers

(defvar *timer-list* nil
  "List of active timers.")

(defstruct timer
  time repeat function args)

(defun run-with-timer (secs repeat function &rest args)
  "Perform an action after a delay of SECS seconds.
Repeat the action every REPEAT seconds, if repeat is non-nil.
SECS and REPEAT may be reals.
The action is to call FUNCTION with arguments ARGS."
  (check-type secs (real 0 *))
  (check-type repeat (or null (real 0 *)))
  (check-type function (or function symbol))
  (let ((timer (make-timer
                :repeat repeat
                :function function
                :args args)))
    (schedule-timer timer secs)
		(setf *timer-list* (merge 'list *timer-list* (list timer) #'< :key #'timer-time))
    timer))

(defun cancel-timer (timer)
  "Remove TIMER from the list of active timers."
  (check-type timer timer)
  (setf *timer-list* (remove timer *timer-list*)))

(defun schedule-timer (timer when)
  (setf (timer-time timer) (+ (get-internal-real-time)
                              (* when internal-time-units-per-second))))

(defun run-expired-timers ()
	(let ((now (get-internal-real-time))
				(timers *timer-list*)
				(pending '())
				(remaining '()))
		(setf *timer-list*
					(dolist (timer timers (sort remaining #'< :key #'timer-time))
						(if (<= (timer-time timer) now)
								(progn (push timer pending)
											 (when (timer-repeat timer)
												 (schedule-timer timer (timer-repeat timer))
												 (push timer remaining)))
								(push timer remaining))))
		(dolist (timer pending)
			(apply (timer-function timer) (timer-args timer)))))

(defun get-next-timeout (timers)
  "Return the number of seconds until the next timeout or nil if there are no timers."
  (when timers
    (max (/ (- (timer-time (car timers)) (get-internal-real-time))
            internal-time-units-per-second)
         0)))

(defun perform-top-level-error-action (c)
  (ecase *top-level-error-action*
    (:message
     (let ((s (format nil "~&Caught '~a' at the top level. Please report this." c)))
       (write-line s)
       (print-backtrace)
       (message "^1*^B~a" s)))
    (:break (invoke-debugger c))
    (:abort
     (throw :top-level (list c (backtrace-string))))))

(defun dswm-internal-loop ()
  "The internal loop that waits for events and handles them."
  (loop
     (run-hook *internal-loop-hook*)
     (handler-bind
         ((xlib:lookup-error (lambda (c)
                               (if (lookup-error-recoverable-p)
                                   (recover-from-lookup-error)
                                   (error c))))
          (warning #'muffle-warning)
          ((or serious-condition error)
           (lambda (c)
             (run-hook *top-level-error-hook*)
             (perform-top-level-error-action c)))
          (t
           (lambda (c)
             ;; some other wacko condition was raised so first try
             ;; what we can to keep going.
             (cond ((find-restart 'muffle-warning)
                    (muffle-warning))
                   ((find-restart 'continue)
                    (continue)))
             ;; and if that fails treat it like a top level error.
             (perform-top-level-error-action c))))
       ;; Note: process-event appears to hang for an unknown
       ;; reason. This is why it is passed a timeout in hopes that
       ;; this will keep it from hanging.
	 (xlib:display-finish-output *display*)
	 (let* ((to (get-next-timeout *timer-list*))
		(timeout (and to (ceiling to)))
		(nevents (xlib:event-listen *display* timeout)))
	   (dformat 10 "timeout: ~a~%" timeout)
	   (when timeout
			 (run-expired-timers))
	   (xlib:with-event-queue (*display*)
				  (when nevents
				    (run-hook *event-processing-hook*)
				    (xlib:process-event *display* :handler #'handle-event :timeout 0))))
	 )))

(defun parse-display-string (display)
  "Parse an X11 DISPLAY string and return the host and display from it."
  (ppcre:register-groups-bind (protocol host ('parse-integer display screen))
			      ("^(?:(.*?)/)?(.*?)?:(\\d+)(?:\\.(\\d+))?" display :sharedp t)
    (values
     ;; clx doesn't like (vector character *)
     (coerce (or host "")
	     '(simple-array character (*)))
     display screen
     (cond (protocol
	    (intern1 protocol :keyword))
	   ((or (string= host "")
		(string-equal host "unix"))
	    :local)
	   (t :internet)))))

(defun dswm-internal (display-str)
  (multiple-value-bind (host display screen protocol) (parse-display-string display-str)
    (declare (ignore screen))
    (setf *display* (xlib:open-display host :display display :protocol protocol)
          (xlib:display-error-handler *display*) 'error-handler)
    (with-simple-restart (quit-dswm "Quit DSWM")
      ;; In the event of an error, we always need to close the display
      (unwind-protect
           (progn
             (let ((*initializing* t))
               ;; we need to do this first because init-screen grabs keys
               (update-modifier-map)
               ;; Initialize all the screens
               (setf *screen-list* (loop for i in (xlib:display-roots *display*)
                                      for n from 0
                                      collect (init-screen i n host)))
               (xlib:display-finish-output *display*)
               ;; Load startup components
               (let ((*package* (find-package *default-package*)))
                 (multiple-value-bind
		     (success err rc)
		     (scratchpad-init)
		   (load-rc-file)
		   (run-with-timer
		    *mode-line-timeout* *mode-line-timeout* 'toggle-mode-line-blink)
		   ;; Switch on mode-line
		   (if (not (head-mode-line (current-head)))
		       (toggle-mode-line (current-screen) (current-head)))
                   (if success
		       (progn
			 (enable-mode-line (current-screen) (current-head) :ds)
			 (first-start))
		     (message "^B^1*Error loading ^b~A^B: ^n~A" rc err))))
               (when *last-unhandled-error*
                 (message-no-timeout "^B^1*DSWM Crashed With An Unhandled Error!~%Copy the error to the clipboard with the 'copy-unhandled-error' command.~%^b~a^B^n~%~%~a"
                          (first *last-unhandled-error*) (second *last-unhandled-error*)))
               (mapc 'process-existing-windows *screen-list*)
               ;; We need to setup each screen with its current window. Go
               ;; through them in reverse so the first screen's frame ends up
               ;; with focus.
               (dolist (s (reverse *screen-list*))
                 ;; map the current group's windows
                 (mapc 'unhide-window (reverse (group-windows (screen-current-group s))))
                 ;; update groups
                 (dolist (g (reverse (screen-groups s)))
                   (dformat 3 "Group windows: ~S~%" (group-windows g))
                   (group-startup g))
                 ;; switch to the (old) current group.
                 (let ((netwm-id (first (xlib:get-property (screen-root s) :_NET_CURRENT_DESKTOP))))
                   (when (and netwm-id (< netwm-id (length (screen-groups s))))
                     (switch-to-group (elt (sort-groups s) netwm-id))))
                 (redraw-current-message (current-screen))))
             ;; Let's manage.
             (let ((*package* (find-package *default-package*)))
               (run-hook *start-hook*)
               (dswm-internal-loop)))
        (xlib:close-display *display*))))
  ;; what should the top level loop do?
  :quit)

;; Usage: (dswm)
(defun dswm (&optional (display-str (or (getenv "DISPLAY") ":0")))
  "Start the deep space window manager."
  (loop
     (let ((ret (catch :top-level
                  (dswm-internal display-str))))
       (setf *last-unhandled-error* nil)
       (cond ((and (consp ret)
                   (typep (first ret) 'condition))
              (format t "~&Caught '~a' at the top level. Please report this.~%~a"
                      (first ret) (second ret))
              (setf *last-unhandled-error* ret))
             ;; we need to jump out of the event loop in order to hup
             ;; the process because otherwise we get errors.
             ((eq ret :hup-process)
                  (apply 'execv (first (argv)) (argv)))
             ((eq ret :restart))
             (t
	      (run-hook *quit-hook*)
              ;; the number is the unix return code
              (return-from dswm 0))))))
